### PR TITLE
authorization: introduce anonymous + agent/service-specific auth

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Introduce the concept of "anonymous access", i.e. what we do for RUM today. Currently anonymous access is restricted to RUM, but we would like to extend this to the iOS agent.

In addition, we would like to be be able to restrict anonymous access to sending  data for a fixed set of service names, to ensure they cannot create arbitrary data streams. To achieve this we change the `authorization.Authorization` interface's `AuthorizedFor` method to accept a new `authorization.Resource` type (replacing the old `elasticsearch.Resource` param) which may contain an agent and/or service name. The authorization implementation can use these to restrict access, either through config or later for API Keys through additional `_has_privilege` queries.

## Checklist

~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~
~- [ ] Documentation has been updated~

## How to test these changes

Non-functional change.

## Related issues

https://github.com/elastic/apm-server/issues/5347

### CODE REVIEW

**Review by simitt — COMMENTED** [review]
Overall changes look good. Left two comments that are no blockers.

File: beater/authprocessor.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/elastic/apm-server/beater/authorization"
+	"github.com/elastic/apm-server/model"
+)
+
+// verifyAuthorizedFor is a model.BatchProcessor that checks authorization
+// for the agent and service name in the metadata.
+func verifyAuthorizedFor(ctx context.Context, meta *model.Metadata) error {
+	result, err := authorization.AuthorizedFor(ctx, authorization.Resource{
+		AgentName:   meta.Service.Agent.Name,
+		ServiceName: meta.Service.Name,
```
[the remark this answers is not in the document]
Comment (axw) [reply beater/authprocessor.go:33]: Yup, this processor is called for every event. Which kinda sucks tbh -- we'll need to be very careful about caching to make sure we don't kill performance. Maybe we can implement some optimisations later, like only call it once for a batch if all are known to have the same (stream-level metadata) service name.

File: beater/api/mux.go [hunk 2]
```diff
@@ -229,7 +231,7 @@ func backendMiddleware(cfg *config.Config, auth *authorization.Handler, m map[re
 	return backendMiddleware
 }
 
-func rumMiddleware(cfg *config.Config, _ *authorization.Handler, m map[request.ResultID]*monitoring.Int) []middleware.Middleware {
+func rumMiddleware(cfg *config.Config, auth *authorization.Handler, m map[request.ResultID]*monitoring.Int) []middleware.Middleware {
```
[the remark this answers is not in the document]
Comment (axw) [reply beater/api/mux.go:234]: Yeah good catch, I started out with some other changes which needed it. I'll revert, thanks.

**Review by simitt — APPROVED** [review]
Changes look good.

**Pull request merged**